File: structure/tree/trie.py

class Trie:
    def __init__(self):
        self.children = [None for i in range(26)]
        self.isEnd = False

    def insert(self, word):
        cur = self
        for c in word:
            if cur.children[ord(c)-ord('a')] is None:
                cur.children[ord(c)-ord('a')] = Trie()
            cur = cur.children[ord(c)-ord('a')]
        cur.isEnd = True

    def search(self, word):
        cur = self
        for c in word:
            if cur.children[ord(c)-ord('a')] is None:
                return False
            cur = cur.children[ord(c)-ord('a')]
        return cur.isEnd

    def starts_with(self, prefix):
        cur = self
        for c in prefix:
            if cur.children[ord(c)-ord('a')] is None:
                return False
            cur = cur.children[ord(c)-ord('a')]
        return True


if __name__ == '__main__':
    trie = Trie()
    trie.insert("hello")
    trie.insert("world")
    print(trie.search("hello"))
    print(trie.search("word"))
    print(trie.starts_with("hell"))

File: structure/tree/trie.cpp
#include <bits/stdc++.h>
using namespace std;


class Trie{

public:
    Trie(): children(), isEnd(false){

    }

    void insert(string word){
        Trie* cur = this;
        for(auto c: word){
            if(cur->children[c-'a'] == nullptr){
                cur->children[c-'a'] = new Trie;
            }
            cur = cur->children[c-'a'];
        }
        cur->isEnd = true;
    }

    bool search(string word){
        Trie* cur = this;
        for(auto c: word){
            if(cur->children[c-'a'] == nullptr) return false;
            cur = cur->children[c-'a'];
        }
        return cur->isEnd; // 注意还要是词的结尾
    }

    // 是否有前缀为prefix的词
    bool startsWith(string prefix){
        Trie* cur = this;
        for(auto c: prefix){
            if(cur->children[c-'a'] == nullptr) return false;
            cur = cur->children[c-'a'];
        }
        return true;
    }


private:
    Trie* children[26]; 
    bool isEnd; // 是否存在单词

};